[Certora] Refactor Certora README (#247)
morpho-org/bundler3: certora/README.md

Reviewer: Note that morpho-blue also has the [folder and file structure section](https://github.com/morpho-org/morpho-blue/tree/main/certora#folder-and-file-structure). Not necessary to keep it, but it would be nice to point to files and rules from when they are mentioned

Author: I think the point of @MathisGD's comment was that we don't really need the folder and file structure here.

Reviewer: Fine to remove as well, but it would be nice to link to some files/rules still, because otherwise it would be difficult to know where they would be from the description alone

Reviewer 2: I think that we could at least have a minimal file structure section, which explains what are the spec, conf, and munging

Author: Done in https://github.com/morpho-org/bundler3/pull/247/commits/1bdbf503427244d6545f0777572583f40fd8ecd1.

The Bundler3 contract enables an EOA to call different endpoint contracts onchain as well as grouping several calls in a single bundle.
These calls may themselves reenter the Bundler3 contract.

## Bundler3 safety restrictions

A key feature of the bundler is to restrict reentering in adapter calls.

> That is we check that reentering an adapter is only possible during a multicall execution.
> We also check that adapters' methods used during a bundle execution may only be called by the Bundler3 contract.
> We check that it's only possible to reenter the adapter using a dedicated Bunlder3 function.

Reviewer: ```suggestion
In particular, it is checked that reentering an adapter is only possible during a multicall execution.
It is also checked that adapters' methods used during a bundle execution may only be called by the Bundler3 contract.
Additionally, the verification makes sure that it's only possible to reenter the adapter using a dedicated Bundler3 function.
```
let's try to avoid the word "we" in general

We also check that adapters' methods used during a bundle execution may only be called by the Bundler3 contract.
We check that it's only possible to reenter the adapter using a dedicated Bunlder3 function.
Bundler3 uses transient storage, it's checked that it's nullified on each entry-point call.
This is checked with a separate configuration as it requires to disable sanity-checks (because `reenter` cannot be an entry-point _per se_).

- [`Bundler3.spec`](specs/Bundler3.spec) checks Bundler3 entry points behave as expected;
- [`AllowancesInvariant.spec`](specs/AllowancesInvariant.spec) check that token allowances are reset;
- [`GeneralAdapter1Reverts.spec`](specs/GeneralAdapter1Reverts.spec) checks that state changes when adapter functions are called, or the execution reverts;

> - [`MorphoZeroConditions.spec`](specs/MorphoZeroConditions.spec) checks that calls to Morpho with zero inputs that revert in Morpho make the adapter revert;

Reviewer: I think the zero condition is missing from the description, but maybe not necessary to add (as it's a bit technical)

> Some adapters may call Morpho entry-points, we check that zero inputs that should revert on Morpho revert directly at the adapter level.

Author: > I think the zero condition is missing from the description, but maybe not necessary to add (as it's a bit technical)

It's here isn't it?

> Throughout the adapters, ERC20 allowances to the adapter may be increased.

Reviewer: ```suggestion
During the execution of an adapter, ERC20 allowance of that adapter to another contract may increase.
```

> We check that such allowances are safe in-between adapters entry-points by resetting them to zero after transactions are performed.

Reviewer: ```suggestion
It is checked that such allowance is reset to zero at the end of the execution.
```

> Note that this not verified for the Paraswap adapter.

Reviewer: ```suggestion
Note that this is not verified for the Paraswap adapter.
```

## General adapter reverts

> In this adapter, we ensure that the state changes when adapter functions are called, or the execution reverts.

Reviewer: ```suggestion
When calling a function of this adapter, it is verified that either the state changes or the execution reverts.
```

> In some cases this property doesn't hold, in those cases a dedicated rule justify this is included.

Reviewer: ```suggestion
This property doesn't hold for every function of the adapters, in those cases a dedicated rule justifying this is included.
```

When calling a function of this adapter, it is verified that either the state changes or the execution reverts.
This property doesn't hold for every function of the adapters, in those cases a dedicated rule justifying this is included.

### Folders and file structure

The [`certora/specs`](specs) folder contains these files:

- [`Bundler3.spec`](specs/Bundler3.spec) describes entry points safety;
- [`AllowancesInvariant.spec`](specs/AllowancesInvariant.spec) describes allowance isolation;
- [`GeneralAdapter1Reverts.spec`](specs/GeneralAdapter1Reverts.spec) describes when state should change or execution reverts;
- [`MorphoZeroConditions.spec`](specs/MorphoZeroConditions.spec) describes how calls to Morpho with zero should behave;
- [`OnlyBundler3.spec`](specs/OnlyBundler3.spec) describes that adapters' methods may only be called by the Bundler3;
- [`ReenterCaller.spec`](specs/ReenterCaller.spec) describes Bundler3 reentering properties;
- [`TransientStorageInvariant.spec`](specs/TransientStorageInvariant.spec) describes the transient storage behavior.

The [`certora/confs`](confs) folder contains a configuration file for each corresponding specification file.

> The [`certora/Makefile`](Makefile) is used to track and perform the required modifications on source files.

Reviewer: There are no `certora/Makefile` in this repo, but instead we should describe the `dispatch` folder

Author: https://github.com/morpho-org/bundler3/pull/247/commits/9011a5c55360f0c465de7316e12413d0cbab6df7
